util/patricia-trie-ethereum/src/rlp_node_codec.rs

```diff
@@ -85,64 +107,61 @@ impl NodeCodec<KeccakHasher> for RlpNodeCodec<KeccakHasher> {
 			// if extension, second item is a node (either SHA3 to be looked up and
 			// fed back into this function or inline RLP which can be fed back into this function).
 			Prototype::List(2) => {
-				let enc_nibble = r.at(0)?.data()?;
-				let from_encoded = if enc_nibble.is_empty() {
-					(NibbleSlice::new(&[]), false)
+				let (partial_rlp, mut partial_offset) = r.at_with_offset(0)?;
+				let partial_payload = partial_rlp.payload_info()?;
+				partial_offset += partial_payload.header_len;
+
+				let (partial, is_leaf) = if partial_rlp.is_empty() {
+					(NibbleSlicePlan::new(partial_offset..partial_offset, 0), false)
 				} else {
+					let partial_header = partial_rlp.data()?[0];
 					// check leaf bit from header.
-					let is_leaf = enc_nibble[0] & 32 == 32;
+					let is_leaf = partial_header & 32 == 32;
 					// Check the header bit to see if we're dealing with an odd partial (only a nibble of header info)
 					// or an even partial (skip a full byte).
-					let (start, byte_offset) = if enc_nibble[0] & 16 == 16 { (0, 1) } else { (1, 0) };
-					(NibbleSlice::new_offset(&enc_nibble[start..], byte_offset), is_leaf)
+					let (start, byte_offset) = if partial_header & 16 == 16 { (0, 1) } else { (1, 0) };
+					let range = (partial_offset + start)..(partial_offset + partial_payload.value_len);
+					(NibbleSlicePlan::new(range, byte_offset), is_leaf)
 				};
-				match from_encoded {
-					(slice, true) => Ok(Node::Leaf(slice, r.at(1)?.data()?)),
-					(slice, false) => Ok(Node::Extension(slice, {
-						let value = r.at(1)?;
-						if value.is_data() && value.size() == KeccakHasher::LENGTH {
-							value.data()?
-						} else {
-							value.as_raw()
-						}
-					})),
-				}
+
+				let (value_rlp, value_offset) = r.at_with_offset(1)?;
+				Ok(if is_leaf {
+					let value = decode_value_range(value_rlp, value_offset)?;
+					NodePlan::Leaf { partial, value }
+				} else {
+					let child = decode_child_handle_plan::<KeccakHasher>(value_rlp, value_offset)?;
+					NodePlan::Extension { partial, child }
+				})
 			},
 			// branch - first 16 are nodes, 17th is a value (or empty).
 			Prototype::List(17) => {
-				let mut nodes = [None as Option<&[u8]>; 16];
+				let mut children = [
+					None, None, None, None, None, None, None, None,
+					None, None, None, None, None, None, None, None,
+				];
 				for i in 0..16 {
```

Comment:
replace with iterator instead i.e, `for (idx, child) in children.iter_mut().enumerate()`